examples/Settings.md
# Enhanced Treemaps
## Settings

```json
{
"type": "enhancedtreemap",
"name": "Enhanced Treemap Settings", "fill": { "s": 0, "l": 0.2 }, "text_size": 18, "halign": "center",
"settings": [
	{ "aspect_ratio": "1:1" },
	{ "fixed_width": true },
	{ "width": 800 },
	{ "sort_by_value": false },
	{ "outer_padding": 6 },
	{ "text_padding": 0 },
	{ "h_text_padding": 7 },
	{ "text_size": 11 },
	{ "h_text_size": 11 },
	{ "text_color": { "l": 0.7 } },
	{ "shading": false },
	{ "h_shading": false },
	{ "shadow": false },
	{ "h_shadow": true },
	{ "h_shadow_size": 3 },
	{ "border_color": { "a": 0 } },
	{ "h_border_color": { "a": 0 } },
	{ "h_fill": { "s": 0.2 } },
	{ "fill": { "a": 0 } }
],
"children": [
	{ "fill": { "a": 0 }, "border_color": { "a": 0 },
		"children": [
			{ "name": "Size & Shape", "text_size": 13, "fill": { "s": 0 }, "children": [ 
				{ "name": "aspect_ratio",   "fill": { "h": 60 }, "children": [ 
					{ "name": "Set the width:height ratio for the treemap. Default=\"1:1\"" } ] },
				{ "name": "fixed_width",    "fill": { "h": 60 }, "children": [ 
					{ "name": "Set whether the treemap has a fixed width (vs automatic width). Default=true" } ] },
				{ "name": "width",          "fill": { "h": 60 }, "children": [ 
					{ "name": "Set the width of the treemap (if it is using a fixed width). Default=800" } ] },
				{ "name": "show_headers",   "fill": { "h": 60 }, "children": [ 
					{ "name": "Set whether headers are shown in the treemap. Default=true" } ] },
				{ "name": "show_values",    "fill": { "h": 60 }, "children": [ 
					{ "name": "Set whether values are shown in the cells. Default=false" } ] },
				{ "name": "sort_by_value",  "fill": { "h": 60 }, "children": [ 
					{ "name": "Set whether cells are sorted by value (vs code block order). Default=true" } ] },
				{ "name": "h_text_size",    "fill": { "h": 120 }, "children": [ 
					{ "name": "Set font size for header text. Default=16" } ] },
				{ "name": "text_size",      "fill": { "h": 180 }, "children": [ 
					{ "name": "Set font size for cell text. Default=13" } ] }
			] },
			{ "name": "Padding", "text_size": 13, "fill": { "s": 0 }, "children": [ 
				{ "name": "outer_padding",  "fill": { "h": 60 }, "children": [ 
					{ "name": "Set the padding around headers and cells. Default=8" } ] },
				{ "name": "h_text_padding", "fill": { "h": 60 }, "children": [ 
					{ "name": "Set the padding around header text. Default=8" } ] },
				{ "name": "text_padding",   "fill": { "h": 180 }, "children": [ 
					{ "name": "Set the padding around cell text. Default=8" } ] }
			] },
			{ "name": "Shading & Shadows", "text_size": 13, "fill": { "s": 0 }, "children": [ 
				{ "name": "h_shading",      "fill": { "h": 120 }, "children": [ 
					{ "name": "Set whether headers have shading. Default=true" } ] },
				{ "name": "shading",        "fill": { "h": 180 }, "children": [ 
					{ "name": "Set whether cells have shading. Default=true" } ] },
				{ "name": "h_shadow",       "fill": { "h": 120 }, "children": [ 
					{ "name": "Set whether headers have drop shadows. Default=true" } ] },
				{ "name": "shadow",         "fill": { "h": 180 }, "children": [ 
					{ "name": "Set whether cells have drop shadows. Default=true" } ] },
				{ "name": "h_shadow_size",  "fill": { "h": 60 }, "children": [ 
					{ "name": "Set size of header drop shadows. Default=4" } ] },
				{ "name": "shadow_size",    "fill": { "h": 60 }, "children": [ 
					{ "name": "Set size of cell drop shadows. Default=4" } ] }
			] },
			{ "name": "Color", "text_size": 13, "fill": { "s": 0 }, "children": [ 
				{ "name": "h_border_color", "fill": { "h": 120 }, "children": [ 
					{ "name": "Set hue, saturation, lightness & alpha for header borders. Default=0,0,0,0.5" } ] },
				{ "name": "border_color",   "fill": { "h": 180 }, "children": [ 
					{ "name": "Set hue, saturation, lightness & alpha for cell borders. Default=0,0,0,0.5" } ] },
				{ "name": "h_fill",         "fill": { "h": 120 }, "children": [ 
					{ "name": "Set hue, saturation, lightness & alpha for filling headers. Default=0,0,0.3,1" } ] },
				{ "name": "fill",           "fill": { "h": 180 }, "children": [ 
					{ "name": "Set hue, saturation, lightness & alpha for filling cells. Default=0,0,0.3,1" } ] },
				{ "name": "h_text_color",   "fill": { "h": 120 }, "children": [ 
					{ "name": "Set hue, saturation, lightness & alpha for header text. Default=0,0,0.9,1" } ] },
				{ "name": "text_color",     "fill": { "h": 180 }, "children": [ 
					{ "name": "Set hue, saturation, lightness & alpha for cell text. Default=0,0,0.8,1" } ] }
			] },
			{ "name": "Alignment", "text_size": 13, "fill": { "s": 0 }, "children": [ 
				{ "name": "h_halign",       "fill": { "h": 120 }, "children": [ 
					{ "name": "Set horizontal alignment for header text. Default=\"left\"" } ] },
				{ "name": "halign",         "fill": { "h": 180 }, "children": [ 
					{ "name": "Set horizontal alignment for cell text. Default=\"left\"" } ] },
				{ "name": "valign",         "fill": { "h": 180 }, "children": [ 
					{ "name": "Set vertical alignment for cell text. Default=\"top\"" } ] }
			] }
		]
	},
	{ "fill": { "a": 0 }, "border_color": { "a": 0 },
		"children": [
			{ "name": "About this Treemap",      "text_size": 13, "fill": { "s": 0, "a": 1 }, "children": [ { "value": 2.25,
				"name": "This treemap shows all available settings for the Enhanced Treemap plugin. Defaults can be changed in the settings panel for the plugin." } ] },
			{ "name": "Gold Settings",  "text_size": 13, "fill": { "h": 60, "s": 0.2, "a": 1 }, "children": [ { "value": 2.25,
				"name": "These settings can be changed in the \"settings\" section of a treemap to override the plugin defaults. They cannot be changed within individual headers or cells." } ] },
			{ "name": "Green Settings", "text_size": 13, "fill": { "h": 120 }, "children": [ { "value": 2.25,
				"name": "These settings can be changed in the \"settings\" section of a treemap to override the plugin defaults. They can also be changed within individual headers to override both the plugin defaults and the treemap settings. Within individual headers remove the \"h_\" prefix." } ] },
			{ "name": "Blue Settings",  "text_size": 13, "fill": { "h": 180 }, "children": [ { "value": 2.25,
				"name": "These settings can be changed in the \"settings\" section of a treemap to override the plugin defaults. They can also be changed within individual cells to override both the plugin defaults and the treemap settings." } ] }
		]
	}
]
}
```



src/settings.ts
import { App, FileSystemAdapter, normalizePath, PluginSettingTab, Setting } from 'obsidian';
import EnhancedTreemapPlugin from './main';

export interface EnhancedTreemapSettings {
    basePath:       string;
    save_as_png:    boolean;
    aspect_w:       number;
    aspect_h:       number;

    // Treemap Settings
    aspect_ratio:   number;
    outer_padding:  number;
    fixed_width:    boolean;
    h_shadow_size:  number;
    h_text_padding: number;
    shadow_size:    number;
    show_headers:   boolean;
    show_values:    boolean;
    sort_by_value:  boolean;
    width:          number;

    // Header Settings
    h_border_color: Array<number>;
    h_fill:         Array<number>;
    h_halign:       string;
    h_shading:      boolean;
    h_shadow:       boolean;
    h_text_color:   Array<number>;
    h_text_size:    number;

    // Cell Settings
    border_color:   Array<number>;
    fill:           Array<number>;
    halign:         string;
    shading:        boolean;
    shadow:         boolean;
    text_color:     Array<number>;
    text_padding:   number;
    text_size:      number;
    valign:         string;

}

export const DEFAULT_SETTINGS: EnhancedTreemapSettings = {
    basePath:       "",
    save_as_png:    true,

    // Treemap Settings
    aspect_ratio:   1,
    aspect_w:       1,
    aspect_h:       1,
    outer_padding:  8,
    fixed_width:    true,
    h_shadow_size:  4,
    h_text_padding: 8,
    shadow_size:    4,
    show_headers:   true,
    show_values:    false,
    sort_by_value:  true,
    width:          800,

    // Header Settings
    h_border_color: [0, 0, 0, 0.5],
    h_fill:         [0, 0, 0.3, 1],
    h_halign:       "left",
    h_shading:      true,
    h_shadow:       true,
    h_text_color:   [0, 0, 0.9, 1],
    h_text_size:    16,

    // Cell Settings
    border_color:   [0, 0, 0, 0.5],
    fill:           [0, 0, 0.3, 1],
    halign:         "left",
    shading:        true,
    shadow:         true,
    text_color:     [0, 0, 0.8, 1],
    text_padding:   8,
    text_size:      13,
    valign:         "top"

}

export class EnhancedTreemapSettingTab extends PluginSettingTab {
    plugin: EnhancedTreemapPlugin;

    constructor(app: App, plugin: EnhancedTreemapPlugin) {
        super(app, plugin);
        this.plugin = plugin;
        var adapter = app.vault.adapter as FileSystemAdapter;
        this.plugin.settings.basePath = normalizePath(adapter.getBasePath());
    }

    // combine hue, saturation, lightness and alpha elements of an array into a string
    getHSLA(setting: Array<number>): string {
        return setting[0] + ", " + setting[1] + ", " + setting[2] + ", " + setting[3];
    }

    // set hue, saturation, lightness and alpha elements of an array from an input string
    setHSLA(value: string, setting: Array<number>): void {
        var values = value.split(", ");
        var h: number = parseFloat(values[0]);
        var s: number = parseFloat(values[1]);
        var l: number = parseFloat(values[2]);
        var a: number = parseFloat(values[3]);
        if (h >= 0 && h <= 360) setting[0] = h;
        if (s >= 0 && s <= 1) setting[1] = s;
        if (l >= 0 && l <= 1) setting[2] = l;
        if (a >= 0 && a <= 1) setting[3] = a;
    }

    // display the plugin settings
    display(): void {
        const { containerEl } = this;

        containerEl.empty();

        containerEl.createEl('h1', { text: 'Enhanced Treemap Settings' } );
        containerEl.createEl('h2', { text: 'Save as PNG' } );

        new Setting(containerEl)
            .setName("Show Save as PNG Button")
            .setDesc("Show Save as PNG Button Above Each Treemap")
            .addToggle((t) => {
                t.setValue(this.plugin.settings.save_as_png).onChange(async (v) => {
                    this.plugin.settings.save_as_png = v;
                    await this.plugin.saveSettings();
                });
            });

        containerEl.createEl('h2', { text: 'Size & Shape' } );

        new Setting(containerEl)
            .setName('Aspect Ratio')
            .setDesc('Default Aspect Ratio (width:height)')
            .addText(text => text
                .setValue(this.plugin.settings.aspect_w.toString() + ":" + this.plugin.settings.aspect_h.toString())
                .onChange(async (value) => {
                    var ratio = value.split(":");
                    var w: number = parseFloat(ratio[0]);
                    var h: number = parseFloat(ratio[1]);
                    if (w > 0 && h > 0) {
                        this.plugin.settings.aspect_w = w;
                        this.plugin.settings.aspect_h = h;
                        this.plugin.settings.aspect_ratio = w / h;
                        await this.plugin.saveSettings();
                    }
                }));

        new Setting(containerEl)
            .setName("Fixed Width")
            .setDesc("Fixed Width or Adpative Width by Default")
            .addToggle((t) => {
                t.setValue(this.plugin.settings.fixed_width).onChange(async (v) => {
                    this.plugin.settings.fixed_width = v;
                    await this.plugin.saveSettings();
                });
            });

        new Setting(containerEl)
            .setName('Treemap Width')
            .setDesc('Default Treemap Width (if Using Fixed Width)')
            .addText(text => text
                .setValue(this.plugin.settings.width.toString())
                .onChange(async (value) => {
                    var w: number = parseFloat(value);
                    if (w > 0) {
                        this.plugin.settings.width = w;
                        await this.plugin.saveSettings();
                    }
                }));

        new Setting(containerEl)
            .setName("Show Headers")
            .setDesc("Show Headers by Default")
            .addToggle((t) => {
                t.setValue(this.plugin.settings.show_headers).onChange(async (v) => {
                    this.plugin.settings.show_headers = v;
                    await this.plugin.saveSettings();
                });
            });

        new Setting(containerEl)
            .setName("Show Values")
            .setDesc("Show Values by Default")
            .addToggle((t) => {
                t.setValue(this.plugin.settings.show_values).onChange(async (v) => {
                    this.plugin.settings.show_values = v;
                    await this.plugin.saveSettings();
                });
            });

        new Setting(containerEl)
            .setName("Sort by Value")
            .setDesc("Sort by Value or No Sort by Default")
            .addToggle((t) => {
                t.setValue(this.plugin.settings.sort_by_value).onChange(async (v) => {
                    this.plugin.settings.sort_by_value = v;
                    await this.plugin.saveSettings();
                });
            });

        new Setting(containerEl)
            .setName('Header Text Size')
            .setDesc('Default Header Text Size')
            .addText(text => text
                .setValue(this.plugin.settings.h_text_size.toString())
                .onChange(async (value) => {
                    var size: number = parseFloat(value);
                    if (size > 0) {
                        this.plugin.settings.h_text_size = size;
                        await this.plugin.saveSettings();
                    }
                }));

        new Setting(containerEl)
            .setName('Cell Text Size')
            .setDesc('Default Cell Text Size')
            .addText(text => text
                .setValue(this.plugin.settings.text_size.toString())
                .onChange(async (value) => {
                    var size: number = parseFloat(value);
                    if (size > 0) {
                        this.plugin.settings.text_size = size;
                        await this.plugin.saveSettings();
                    }
                }));


        containerEl.createEl('br');
        containerEl.createEl('h2', {text: 'Shading & Shadows'});

        new Setting(containerEl)
            .setName('Header Shadow Size')
            .setDesc('Default Header Shadow Size')
            .addText(text => text
                .setValue(this.plugin.settings.h_shadow_size.toString())
                .onChange(async (value) => {
                    var size: number = parseFloat(value);
                    if (size > 0) {
                        this.plugin.settings.h_shadow_size = size;
                        await this.plugin.saveSettings();
                    }
                }));

        new Setting(containerEl)
            .setName('Cell Shadow Size')
            .setDesc('Default Cell Shadow Size')
            .addText(text => text
                .setValue(this.plugin.settings.shadow_size.toString())
                .onChange(async (value) => {
                    var size: number = parseFloat(value);
                    if (size > 0) {
                        this.plugin.settings.shadow_size = size;
                        await this.plugin.saveSettings();
                    }
                }));

        new Setting(containerEl)
            .setName("Header Shading")
            .setDesc("Add Header Shading by Default")
            .addToggle((t) => {
                t.setValue(this.plugin.settings.h_shading).onChange(async (v) => {
                    this.plugin.settings.h_shading = v;
                    await this.plugin.saveSettings();
                });
            });

        new Setting(containerEl)
            .setName("Cell Shading")
            .setDesc("Add Cell Shading by Default")
            .addToggle((t) => {
                t.setValue(this.plugin.settings.shading).onChange(async (v) => {
                    this.plugin.settings.shading = v;
                    await this.plugin.saveSettings();
                });
            });

        new Setting(containerEl)
            .setName("Header Shadows")
            .setDesc("Add Drop Shadows to Headers by Default")
            .addToggle((t) => {
                t.setValue(this.plugin.settings.h_shadow).onChange(async (v) => {
                    this.plugin.settings.h_shadow = v;
                    await this.plugin.saveSettings();
                });
            });

        new Setting(containerEl)
            .setName("Cell Shadows")
            .setDesc("Add Cell Shadows by Default")
            .addToggle((t) => {
                t.setValue(this.plugin.settings.shadow).onChange(async (v) => {
                    this.plugin.settings.shadow = v;
                    await this.plugin.saveSettings();
                });
            });


        containerEl.createEl('br');
        containerEl.createEl('h2', { text: 'Color' } );

        new Setting(containerEl)
            .setName("Header Border Color")
            .setDesc("Default Hue, Saturation, Lightness and Alpha (Transparency) for Header Borders")
            .addText((text) => {
                var hsla = this.getHSLA(this.plugin.settings.h_border_color);
                text.setValue(hsla).onChange(async (value) => {
                    this.setHSLA(value, this.plugin.settings.h_border_color);
                    await this.plugin.saveSettings();
                });
            });

        new Setting(containerEl)
            .setName("Cell Border Color")
            .setDesc("Default Hue, Saturation, Lightness and Alpha (Transparency) for Cell Borders")
            .addText((text) => {
                var hsla = this.getHSLA(this.plugin.settings.border_color);
                text.setValue(hsla).onChange(async (value) => {
                    this.setHSLA(value, this.plugin.settings.border_color);
                    await this.plugin.saveSettings();
                });
            });

        new Setting(containerEl)
            .setName("Header Fill Color")
            .setDesc("Default Hue, Saturation, Lightness and Alpha (Transparency) for Filling Headers")
            .addText((text) => {
                var hsla = this.getHSLA(this.plugin.settings.h_fill);
                text.setValue(hsla).onChange(async (value) => {
                    this.setHSLA(value, this.plugin.settings.h_fill);
                    await this.plugin.saveSettings();
                });
            });

        new Setting(containerEl)
            .setName("Cell Fill Color")
            .setDesc("Default Hue, Saturation, Lightness and Alpha (Transparency) for Filling Cells")
            .addText((text) => {
                var hsla = this.getHSLA(this.plugin.settings.fill);
                text.setValue(hsla).onChange(async (value) => {
                    this.setHSLA(value, this.plugin.settings.fill);
                    await this.plugin.saveSettings();
                });
            });

        new Setting(containerEl)
            .setName("Header Text Color")
            .setDesc("Default Hue, Saturation, Lightness and Alpha (Transparency) for Header Text")
            .addText((text) => {
                var hsla = this.getHSLA(this.plugin.settings.h_text_color);
                text.setValue(hsla).onChange(async (value) => {
                    this.setHSLA(value, this.plugin.settings.h_text_color);
                    await this.plugin.saveSettings();
                });
            });

        new Setting(containerEl)
            .setName("Cell Text Color")
            .setDesc("Default Hue, Saturation, Lightness and Alpha (Transparency) for Cell Text")
            .addText((text) => {
                var hsla = this.getHSLA(this.plugin.settings.text_color);
                text.setValue(hsla).onChange(async (value) => {
                    this.setHSLA(value, this.plugin.settings.text_color);
                    await this.plugin.saveSettings();
                });
            });


        containerEl.createEl('br');
        containerEl.createEl('h2', { text: 'Padding' } );

        new Setting(containerEl)
            .setName('Cell Padding')
            .setDesc('Default Padding Size Around Cells')
            .addText(text => text
                .setValue(this.plugin.settings.outer_padding.toString())
                .onChange(async (value) => {
                    var padding: number = parseFloat(value);
                    if (padding > 0) {
                        this.plugin.settings.outer_padding = padding;
                        await this.plugin.saveSettings();
                    }
                }));

        new Setting(containerEl)
            .setName('Header Text Padding')
            .setDesc('Default Padding Size Around Header Text')
            .addText(text => text
                .setValue(this.plugin.settings.h_text_padding.toString())
                .onChange(async (value) => {
                    var padding: number = parseFloat(value);
                    if (padding > 0) {
                        this.plugin.settings.h_text_padding = padding;
                        await this.plugin.saveSettings();
                    }
                }));

        new Setting(containerEl)
            .setName('Cell Text Padding')
            .setDesc('Default Padding Size Around Cell Text')
            .addText(text => text
                .setValue(this.plugin.settings.text_padding.toString())
                .onChange(async (value) => {
                    var padding: number = parseFloat(value);
                    if (padding > 0) {
                        this.plugin.settings.text_padding = padding;
                        await this.plugin.saveSettings();
                    }
                }));

        containerEl.createEl('br');
        containerEl.createEl('h2', { text: 'Alignment' } );

        new Setting(containerEl)
            .setName("Header Horizontal Alignment")
            .setDesc("Default Horizontal Alignment for Header Text")
            .addDropdown((d) => {
                d.addOption("left", "Left");
                d.addOption("center", "Center");
                d.addOption("right", "Right");
                d.setValue(this.plugin.settings.h_halign);
                d.onChange(async (v) => {
                    this.plugin.settings.h_halign = d.getValue() as string;
                    await this.plugin.saveSettings();
                });
            });

        new Setting(containerEl)
            .setName("Cell Horizontal Alignment")
            .setDesc("Default Cell Text Horizontal Alignment")
            .addDropdown((d) => {
                d.addOption("left", "Left");
                d.addOption("center", "Center");
                d.addOption("right", "Right");
                d.setValue(this.plugin.settings.halign);
                d.onChange(async (v) => {
                    this.plugin.settings.halign = d.getValue() as string;
                    await this.plugin.saveSettings();
                });
            });

        new Setting(containerEl)
            .setName("Cell Vertical Alignment")
            .setDesc("Default Cell Text Vertical Alignment")
            .addDropdown((d) => {
                d.addOption("top", "Top");
                d.addOption("center", "Center");
                d.addOption("bottom", "Bottom");
                d.setValue(this.plugin.settings.valign);
                d.onChange(async (v) => {
                    this.plugin.settings.valign = d.getValue() as string;
                    await this.plugin.saveSettings();
                });
            });

    }
}

